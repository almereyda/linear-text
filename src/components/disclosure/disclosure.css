.disclosure__summary {
  /* Hide the drop-down triangle indicator. */
  list-style: none;
}

/* Hack: copied from button:active to avoid flicker when changing active
   states. */
.disclosure[open] > .disclosure__summary {
  border-style: inset;
}
