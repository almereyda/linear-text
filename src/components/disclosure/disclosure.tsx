import type {ReactNode} from 'react'

import {useBool} from '../../hooks/use-bool'
import {useCallback, useMemo} from 'react'

import './disclosure.css'

export type DisclosureProps = Readonly<{
  children: ReactNode
  'data-testid'?: string
  onToggle?(open: boolean): void
  open?: boolean
  summary: ReactNode
  title: string
}>

export function Disclosure({
  children,
  'data-testid': testID,
  onToggle,
  open: initOpen = false,
  summary,
  title
}: DisclosureProps) {
  const {val: open, toggle: toggleOpen} = useBool(initOpen)

  const onToggleCb = useCallback(() => {
    toggleOpen()
    onToggle?.(!open)
  }, [onToggle, open, toggleOpen])

  const summaryClassName = useMemo(
    () => `disclosure__summary buttonish ${open ? 'buttonish--active' : ''}`,
    [open]
  )

  return (
    <details
      className='disclosure'
      data-testid={testID}
      onToggle={onToggleCb}
      open={open}
    >
      <summary className={summaryClassName} data-title={title} role='button'>
        {summary}
      </summary>
      {children}
    </details>
  )
}
