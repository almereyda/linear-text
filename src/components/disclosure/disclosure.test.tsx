import {Disclosure} from './disclosure'
import {fireEvent, render, screen} from '@testing-library/react'

test('matches snapshot', () => {
  render(
    <Disclosure data-testid='disclosure' summary='Summary'>
      Children
    </Disclosure>
  )
  const disclosure = screen.getByTestId('disclosure')
  expect(disclosure).toMatchSnapshot()
})

test('renders summary and children', () => {
  render(
    <Disclosure data-testid='disclosure' open summary='Summary'>
      Children
    </Disclosure>
  )
  const disclosure = screen.getByTestId('disclosure')
  expect(disclosure).toHaveTextContent('SummaryChildren')
})

test('reports toggle', () => {
  const onToggle = jest.fn()
  render(
    <Disclosure data-testid='disclosure' onToggle={onToggle} summary='Summary'>
      Children
    </Disclosure>
  )

  const disclosure = screen.getByTestId('disclosure')
  fireEvent(disclosure, new Event('toggle'))
  expect(onToggle).toHaveBeenCalledTimes(1)
})
