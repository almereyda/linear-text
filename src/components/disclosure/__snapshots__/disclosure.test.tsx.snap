// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<details
  class="disclosure"
  data-testid="disclosure"
>
  <summary
    class="disclosure__summary buttonish "
    role="button"
  >
    Summary
  </summary>
  Children
</details>
`;
